### PR TITLE
feat: Collecting data from API registration workflow

### LINKED ISSUE #113: Collecting data from API registration workflow
We need to store the contact data collected in the form that users see [when they register for an API account](https://www.figma.com/file/LZVQecGplb4Lnuh7a24Jef/GTFS-Tool-Forms?type=design&node-id=165-95&mode=dev).

### PR DESCRIPTION
**Pull request by cka-y**
**Summary:**

We need to store the contact data collected in the form that users see [when they register for an API account](https://www.figma.com/file/LZVQecGplb4Lnuh7a24Jef/GTFS-Tool-Forms?type=design&node-id=165-95&mode=dev).

Closes #113 
image

**Expected behavior:** 
- Before the user registers: the registration page is displayed 
- Once the user is registered: should be redirected to 

**Testing tips:**
Using any sign up/login method follow the following steps:
- Sign up 
- Complete the registration process (you should be redirected to the registration page)
- Upon completion you should be redirected to the Account page (no action required on your part)
- Visit [the datastore console](https://console.cloud.google.com/datastore/databases/-default-/entities;kind=web_api_users;ns=__$DEFAULT$__/query/kind?project=mobility-feeds-dev&lastRefresh=1701276748158) to see that your information was registered.
- Sign out
- Sign in again (you should directly be redirected to the account page with your registration information visible)

*Note 1*: If you sign in with github and your email is hidden, you should not have access to the option to register to email announcements from the registration page
*Note 2*: If you register to email announcements, you should be able to see it in your account page as displayed bellow.
image

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `./scripts/api-tests.sh` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Review by qcdyx — APPROVED** [review]
LGTM

File: functions/user-api/src/impl/user-api-impl.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+export const retrieveUser = async (
+  datastore: Datastore,
+  uid: string):
+  Promise<Entity | undefined> => {
+  /**
+   * Retrieve the user information from datastore
+   * @param datastore - Datastore object
+   * @param uid - The user's uid
+   */
```
Comment (davidgamez) [inline functions/user-api/src/impl/user-api-impl.ts:16]: Applicable to all function comments, add the comment block outside the function definition.

File: functions/user-api/src/impl/user-api-impl.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        },
+        {
+          name: "registrationCompletionTime",
+          value: new Date().toJSON(),
+        },
+      ],
+    };
+    await datastore.save(entity).catch((error) => {
+      throw new HttpsError("unknown", error.message);
+    });
```
Comment (davidgamez) [inline functions/user-api/src/impl/user-api-impl.ts:93]: Let's add a unit test to this section to verify that an exception is propagated. I think the exception will not propagate to the main function as it's thrown in a different context, the created promise context. I would replace this block with a try/catch like:
```suggestion
   try {
    await datastore.save(entity);
   } catch((error) => {
       throw new HttpsError("unknown", error.message);
   })
```

File: functions/firebase.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        "node_modules",
+        ".git",
+        "__tests__",
+        "firebase-debug.log",
+        "firebase-debug.*.log"
+      ],
+      "predeploy": [
+        "yarn --cwd \"$RESOURCE_DIR\" install",
+        "yarn --cwd \"$RESOURCE_DIR\" lint",
+        "yarn --cwd \"$RESOURCE_DIR\" run test",
```
[the remark this answers is not in the document]
Comment (cka-y) [reply functions/firebase.json:16]: No since there is no package.json within the `functions` subfolder

[the remark this answers is not in the document]
Comment (davidgamez) [reply functions/firebase.json:16]: I see; my concern is that the test results are not visible enough to debug when failures happen on GH. However, we can leave this now and get back to it when we add more  functions.

[the remark this answers is not in the document]
Comment (cka-y) [reply functions/firebase.json:16]: Currently working on a potential fix...

**Review by davidgamez — APPROVED** [review]
Great work!

**Pull request merged**